src/v/redpanda/admin_server.cc

```diff
@@ -991,7 +991,25 @@ void admin_server::register_security_routes() {
           auto err
             = co_await _controller->get_security_frontend().local().create_user(
               username, credential, model::timeout_clock::now() + 5s);
-          vlog(logger.debug, "Creating user {}:{}", err, err.message());
+          vlog(
+            logger.debug,
+            "Creating user '{}' {}:{}",
+            username,
+            err,
+            err.message());
+
+          if (err == cluster::make_error_code(cluster::errc::user_exists)) {
+              // Idempotency: if user is same as one that already exists,
+              // suppress the user_exists error and return success.
```

Comment:
Is this the preferred way to handle this? I'm not sure if I'd rather see "user exists" or "success" if trying to recreate a user with the same password. Offhand I think I'd expect an "exists" error if I create a user twice in a row, similar to how creating a topic twice returns an exists error the second time.

We can't really compare to Kafka here through the protocol because the only command is `AlterUserScramCredentials`, so an upsert is always a success. I'm not sure how the zookeeper api works.

Reply:
The motivation for doing the idempotence server side is:
- A retry loop needs to exist for any automated setup process that creates users, whether it's done client or server side.  Otherwise the client has a high chance of failing, when it tries to create a user while the initial controller leader election is happening.
- HTTP doesn't have a clear response code for indicating "already exists and is identical" -- the closest is 409, but that just says the resource couldn't be created, not that the existing resource was identical.
- Clients can't get at credentials, so a client has no way of distinguishing (other than attempting to log in) between the case where an identical user exists and a case where the username is already taken by a user with identical credentials.

It is to some extent a matter of API taste though.